refactor 3636 nnnnat accounts revisions hooks events

nnnnat opened the pull request:
Fixes #3636 

## Summary
Replaced all Account & Revision `.before`, `.after` collection hooks to use `Hooks.Events` API.
- Updated `Revisons.after.update(callback)` to be `Hooks.Events.add("afterRevisionUpdate", callback)` and added `Hooks.Events.run("afterRevisionUpdate", userId, revision)` after every `Revisons.update(...)` call.
- Updated `Accounts.after.insert(callback)` to be `Hooks.Events.add("afterAccountsInsert", callback)` and added a `Hooks.Events.run("afterAccountsInsert", userId, user)` after every `Accounts.insert(...)` call.
- Updated `Accounts.after.remove(callback)` to be `Hooks.Events.add("afterAccountsRemove", callback)` and added a `Hooks.Events.run("afterAccountsRemove", userId, user)` after every `Accounts.remove(...)` call.
- Updated `Accounts.after.update(callback)` to be `Hooks.Events.add("afterAccountsUpdate", callback)` and added a `Hooks.Events.run("afterAccountsUpdate", userId, user)` after every `Accounts.update(...)` call.
- Removed `.direct` from any `Accounts` or `Revisions` collection calls

## Testing
- Run `reaction test` to verify all test still pass
- Run `npm run lint` to verify no code style errors
- After running `reaction reset` login as an Admin and set up payments/shipping.
- As an admin make some changes to the example product and publish.
- In another browser/window purchase the example product as a guest.
- In another browser/window register as a new user and purchase the example product.

## Notes
- I'm not super confident I'm passing the appropriate params in every `Hooks.Events.run(...)` call. I'll try to call them out with inline comments in this PR.
- I had a few functions in `server/methods/accounts/accounts.js` that were returning `Accounts.update(...)`, not sure if my solution is correct.
- Had one instance of `Accounts.update` in a React component, I tried to add the `Hooks.Events.run("afterAccountsUpdate")` but I would run into an error in the browser console about trying to load server code in the client. Not sure how to best deal with this or if I can just leave it as is?

aldeed commented on imports/plugins/core/i18n/client/containers/currencyDropdown.js:
```diff
@@ -16,6 +16,7 @@ const handlers = {
     // UserProfile and ShopMembers publications.
     //
     Accounts.update(Meteor.userId(), { $set: { "profile.currency": currencyName } });
+    // Hooks.Events.run("afterAccountUpdate", Meteor.userId(), Meteor.user());
```
The correct fix here is to create a new server method in `server/methods/accounts/accounts.js` that does `Accounts.update` followed by running the hook. It could be called something like `"accounts/setProfileCurrency"`. It does not need any security on it other than making sure `this.userId` is set (caller is logged in) and then using `this.userId` for the `Accounts.update` ID.

nnnnat commented:
@aldeed  I've updated this PR with the requested changes
- Updated Account hooks to take two id's as params instead of a `userId` and `doc` and updated everyplace I'm using the hook.
- Created a `accounts/setProfileCurrency` method that updates the Accounts currency and call the Hooks event on the next line
- Replaced the `Accounts.update(...)` with the new `Meteor.call("accounts/setProfileCurrency")` method in the currencyDropdown component
- Retested everything

aldeed commented on imports/plugins/core/revisions/server/hooks.js:
```diff
@@ -453,6 +454,7 @@ Products.before.update(function (userId, product, fieldNames, modifier, options)
                 "documentData.price": modifier.$set.price
               }
             });
+            Hooks.Events.run("afterRevisionsUpdate", userId, productRevision);
```
Again, should be the updated revision doc passed to the hook. Either extend to simulate the `$set` above, or do a `findOne()` to get it out of the DB if there isn't a good way to simulate.

aldeed commented on imports/plugins/core/revisions/server/hooks.js:
```diff
@@ -464,6 +466,7 @@ Products.before.update(function (userId, product, fieldNames, modifier, options)
                 "documentData.isVisible": modifier.$set.isVisible
               }
             });
+            Hooks.Events.run("afterRevisionsUpdate", userId, productRevision);
```
Again, should be the updated revision doc passed to the hook. Either extend to simulate the `$set` above, or do a `findOne()` to get it out of the DB if there isn't a good way to simulate.

aldeed commented on imports/plugins/core/revisions/server/hooks.js:
```diff
@@ -537,6 +540,7 @@ Products.before.update(function (userId, product, fieldNames, modifier, options)
   }
 
   Revisions.update(revisionSelector, revisionModifier);
+  Hooks.Events.run("afterRevisionsUpdate", userId, productRevision);
```
Again, should be the updated revision doc passed to the hook. Either extend to simulate the `$set` above, or do a `findOne()` to get it out of the DB if there isn't a good way to simulate.

aldeed commented on imports/plugins/core/revisions/server/hooks.js:
```diff
@@ -643,6 +647,7 @@ Products.before.remove((userId, product) => {
       "workflow.status": "revision/remove"
     }
   });
+  Hooks.Events.run("afterRevisionsUpdate", userId, productRevision);
```
Again, should be the updated revision doc passed to the hook. Either extend to simulate the `$set` above, or do a `findOne()` to get it out of the DB if there isn't a good way to simulate.

aldeed commented on imports/plugins/included/connectors-shopify/server/methods/import/customers.js:
```diff
@@ -166,9 +166,11 @@ export const methods = {
 
             // Insert customer, save id
             const reactionCustomerId = Accounts.insert(reactionCustomer, { publish: true });
+            Hooks.Events.run("afterAccountsInsert", reactionCustomerId, reactionCustomerId);
```
The first arg should be `Meteor.userId()` instead of `reactionCustomerId`. It is supposed to be the user who is triggering the operation, i.e. user who called the server method.

aldeed commented on imports/plugins/included/connectors-shopify/server/methods/import/customers.js:
```diff
@@ -166,9 +166,11 @@ export const methods = {
             ids.push(reactionCustomerId);
 
             Accounts.update({ _id: reactionCustomerId }, { publish: true });
+            Hooks.Events.run("afterAccountsUpdate", reactionCustomerId, reactionCustomerId);
```
The first arg should be `Meteor.userId()`

aldeed commented on imports/plugins/included/jobcontrol/server/jobs/cart.js:
```diff
@@ -60,6 +60,7 @@ export function cartCleanupJob() {
             _id: cart.userId,
             emails: []
           });
+          Hooks.Events.run("afterAccountsRemove", cart.userId, user._id);
```
The first arg should be `null` because this is a background job, which does not run with a user context.

aldeed commented on server/methods/accounts/accounts.js:
```diff
@@ -63,6 +63,7 @@ export function verifyAccount(email, token) {
           "emails.$.verified": true
         }
       });
+      Hooks.Events.run("afterAccountsUpdate", account._id, account._id);
```
The first arg should be `Meteor.userId()`

aldeed commented on server/methods/accounts/accounts.js:
```diff
@@ -366,6 +368,7 @@ export function addressBookAdd(address, accountUserId) {
           "profile.addressBook.$.isShippingDefault": false
         }
       });
+      Hooks.Events.run("afterAccountsUpdate", userId, account._id);
```
The first arg should be `Meteor.userId()`

aldeed commented on server/methods/accounts/accounts.js:
```diff
@@ -376,6 +379,7 @@ export function addressBookAdd(address, accountUserId) {
           "profile.addressBook.$.isBillingDefault": false
         }
       });
+      Hooks.Events.run("afterAccountsUpdate", userId, account._id);
```
The first arg should be `Meteor.userId()`

aldeed commented on server/methods/accounts/accounts.js:
```diff
@@ -468,6 +472,7 @@ export function addressBookUpdate(address, accountUserId, type) {
             "profile.addressBook.$.isShippingDefault": false
           }
         });
+        Hooks.Events.run("afterAccountsUpdate", userId, account._id);
```
The first arg should be `Meteor.userId()`

aldeed commented on server/methods/accounts/accounts.js:
```diff
@@ -490,6 +495,7 @@ export function addressBookUpdate(address, accountUserId, type) {
             "profile.addressBook.$.isBillingDefault": false
           }
         });
+        Hooks.Events.run("afterAccountsUpdate", userId, account._id);
```
The first arg should be `Meteor.userId()`

aldeed commented on server/methods/accounts/accounts.js:
```diff
@@ -518,10 +524,12 @@ export function addressBookUpdate(address, accountUserId, type) {
     userId,
     "profile.addressBook._id": address._id
   }, accountsUpdateQuery);
+  Hooks.Events.run("afterAccountsUpdate", userId, account._id);
```
The first arg should be `Meteor.userId()`

aldeed commented on server/methods/accounts/accounts.js:
```diff
@@ -560,6 +569,8 @@ export function addressBookRemove(addressId, accountUserId) {
       }
     }
   });
+  Hooks.Events.run("afterAccountsUpdate", userId, account._id);
```
The first arg should be `Meteor.userId()`

aldeed commented on server/methods/core/groups.js:
```diff
@@ -178,7 +178,7 @@ Meteor.methods({
       setUserPermissions({ _id: userId }, permissions, shopId);
       Accounts.update({ _id: userId }, { $set: { groups: newGroups } });
-
+      Hooks.Events.run("afterAccountsUpdate", userId, userId);
```
First arg should be `loggedInUserId`

aldeed commented on server/methods/core/groups.js:
```diff
@@ -226,6 +226,7 @@ Meteor.methods({
     try {
       setUserPermissions(user, defaultCustomerGroupForShop.permissions, shopId);
       Accounts.update({ _id: userId, groups: groupId }, { $set: { "groups.$": defaultCustomerGroupForShop._id } }); // replace the old id with new id
+      Hooks.Events.run("afterAccountsUpdate", userId, userId);
```
First arg should be `Meteor.userId()`

aldeed commented on server/methods/core/groups.js:
```diff
@@ -272,4 +273,5 @@ function setUserPermissions(users, permissions, shopId) {
 Hooks.Events.add("afterCreateDefaultAdminUser", (user) => {
   const group = Groups.findOne({ slug: "owner", shopId: Reaction.getShopId() });
   Accounts.update({ _id: user._id  }, { $set: { groups: [group._id] } });
+  Hooks.Events.run("afterAccountsUpdate", user._id, user._id);
```
First arg should be `null` since this runs without a user context

aldeed commented on server/methods/core/shop.js:
```diff
@@ -141,7 +141,7 @@ Meteor.methods({
       }
     });
-
+    Hooks.Events.run("afterAccountsUpdate", currentUser.userId, shopUser.userId);
```
Wouldn't it be `currentUser._id` and `shopUser._id` instead of `.userId`?

aldeed commented on server/methods/core/shop.js:
```diff
@@ -259,6 +259,7 @@ Meteor.methods({
       }
 
       Collections.Accounts.update(user._id, { $set: { "profile.currency": profileCurrency } });
+      Hooks.Events.run("afterAccountsUpdate", user._id, user._id);
```
You could replace these two lines with your new method that already calls the hook

```js
Meteor.call("accounts/setProfileCurrency", profileCurrency);
```

aldeed commented on imports/plugins/core/revisions/server/hooks.js:
```diff
@@ -537,6 +552,10 @@ Products.before.update(function (userId, product, fieldNames, modifier, options)
+  Hooks.Events.run("afterRevisionsUpdate", userId, {
+    ...productRevision,
+    workflow: { ...productRevision.workflow, status: "revisions/update" }
+  });
```
This one may have a bunch of different changes in `revisionModifier` after going through the above, so unfortunately probably the best choice here is to `findOne` after the update.

nnnnat commented:
@aldeed this PR should be good now.

- Updated all revision hooks to pass the correct revision params
- Fixed first param issues on all Accounts Hooks
- Replaced the Accounts.update() + Hooks calls to set account currency in server/methods/core/shop.js, with the new "accounts/setProfileCurrency" method
- Added a `check(currecnyName, String)` to the new "accounts/setProfileCurrency" method to suppress a server warning.
- Retested
